// @umijs/preset-built-in/lib/plugins/commands/generate/PageGenerator/createPageGenerator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _path() {
  const data = require("path");

  _path = function _path() {
    return data;
  };

  return data;
}

function _utils() {
  const data = require("@umijs/utils");

  _utils = function _utils() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _default({
  api
}) {
  return class PageGenerator extends _utils().Generator {
    constructor(opts) {
      super(opts);
    }

    writing() {
      var _this = this;

      return _asyncToGenerator(function* () {
        const _this$args$_ = _slicedToArray(_this.args._, 1),
              path = _this$args$_[0];

        const jsExt = _this.args.typescript ? '.tsx' : '.js';
        const cssExt = _this.args.less ? '.less' : '.css';

        _this.copyTpl({
          templatePath: (0, _path().join)(__dirname, `page${jsExt}.tpl`),
          target: (0, _path().join)(api.paths.absPagesPath, `${path}${jsExt}`),
          context: {
            path,
            name: (0, _path().basename)(path),
            cssExt
          }
        });

        _this.copyTpl({
          templatePath: (0, _path().join)(__dirname, `page.css.tpl`),
          target: (0, _path().join)(api.paths.absPagesPath, `${path}${cssExt}`),
          context: {
            color: (0, _utils().randomColor)()
          }
        });
      })();
    }

  };
}
